### DISCUSSION TITLE
Graph attributes

### DISCUSSION THREAD

**Discussion by nichtich** [pg-format/pg-formatter#31] [Ideas]
Some formats allow for graph attributes and similar metadata associated (e.g. schemas) with a graph. This could be supported in PG format with a node identifier `:` like this

~~~
: created: 2022
  modified: 2024
  title: "example graph"
~~~

In PG-JSON(L) this would translate to an additional entry without `id`, `from`, `to`:

~~~
{"labels":[],"properties":{"created":[2022],"modified":[2024],"title":["example graph"]}}
~~~
 
Semantics of the graph metadata should be out of the scope of PG format and data model.

My use case is a custom encoding/mapping of RDF in PG:

~~~
: ns:bibo: http://purl.org/ontology/bibo/
: ns:dc: http://purl.org/dc/elements/1.1/
: ns:foaf: http://xmlns.com/foaf/0.1/

http://example.org/person/Mark_Twain :foaf:person
http://example.org/books/Tom_Sawyer :bibo:Book
  dc:title "The_Adventures_of_Tom_Sawyer"

http://example.org/books/Tom_Sawyer -> http://example.org/person/Mark_Twain :dc:creator
~~~

**Comment by nichtich** [comment]
I have collected the idea at https://github.com/pg-format/pg-formatter/wiki/Extensions and would prefer to first finish a stable version 1.0.0 before adding new features.

**Comment by hchiba1** [comment]
Interesting idea, but there might be other ways to add graph attributes (metadata).
For example, how about the following _*doc_ style comments:
```
# @title example graph
# @created 2022
# @modified 2024
```

**Comment by hchiba1** [comment]
By considering such issue is separated from the PG-model and should be treated by application layer,
it might also be possible to write as:
```turtle
@prefix bibo: http://purl.org/ontology/bibo/
@prefix dc: http://purl.org/dc/elements/1.1/
@prefix foaf: http://xmlns.com/foaf/0.1/

http://example.org/person/Mark_Twain : foaf:person
http://example.org/books/Tom_Sawyer : bibo:Book
  dc:title : "The Adventures of Tom Sawyer"

http://example.org/books/Tom_Sawyer -> http://example.org/person/Mark_Twain : dc:creator
```
How to interpret the specific node with ID `@prefix` is up to the application layer.

**Reply by nichtich** [reply]
If graph attributes are not part of the PG data model, we should better not recommend any way to encode them. [Some graph formats support graph attributes](https://github.com/pg-format/pgraphs/blob/dev/docs/features.csv), so these cannot be exchanged with PG formats If graph attributes are going to be added, both PG data model and serializations (PG format, PG-JSON(L)) must be extended. There are several ways to extend PG format for this purpose and I'd like to have graph attributes in PG but would prefer to first finish the rest of the specification.
